aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -387,32 +487,57 @@ Tensor& add_out_dense_sparse_cpu(Tensor& r, const Tensor& dense, const SparseTen
     dense.sizes(), " while other has size ", sparse_.sizes(), " (FYI: dense-sparse addition does not currently support broadcasting)");
 
   r.resize_as_(dense);
-  SparseTensor sparse = sparse_.coalesce();
-
-  LongTensor indices = sparse._indices();
-  Tensor values = sparse._values();
-  int64_t nDim = dense.dim();
-  int64_t nDimI = sparse.sparse_dim();
-
   if (!is_same_tensor(r, dense)) r.copy_(dense);
-  if (sparse._nnz() == 0) return r;
 
-  // accessors rely on nnz test
-  if (nDim > nDimI) {
-    auto indices_accessor = indices.accessor<int64_t, 2>();
-    for (int64_t k = 0; k < sparse._nnz(); k++) {
-      Tensor dstBuffer = r;
-      for (int64_t d = 0; d < sparse.sparse_dim(); d++) {
-        dstBuffer = dstBuffer.select(0, indices_accessor[d][k]);
-      }
-      Tensor srcBuffer = values.select(0, k);
-      dstBuffer.add_(srcBuffer, value);
+  auto sparse_nnz = sparse_._nnz();
+  if (sparse_nnz == 0) return r;
+
+  int64_t dense_dim = dense.dim();
+  int64_t sparse_dim = sparse_.sparse_dim();
+  Tensor values = sparse_._values();
+  auto values_type = values.scalar_type();
+  bool sparse_is_coalesced = (sparse_.is_coalesced() || sparse_nnz < 2);
+  bool result_is_contiguous =((r.storage().data() != nullptr) && r.is_contiguous());
+  bool value_is_contiguous =((values.storage().data() != nullptr) && values.is_contiguous());
+  bool is_contiguous =  (result_is_contiguous && value_is_contiguous);
+
+  LongTensor indices;
+  if (is_contiguous && sparse_is_coalesced) {
+    indices = sparse_._indices();
```

Comment:
You are doing it in every branch, why not replace line above with `LongTensor indices = sparse_._indices();`

Reply:
The branch for slow non-contiguous sparse tensor will call "SparseTensor sparse = sparse_.coalesce();" first then get the "indices = sparse._indices();", a little difference with other two faster branch